include/pybind11/pybind11.h

```diff
@@ -427,13 +434,16 @@ class cpp_function : public function {
             if (it->doc && strlen(it->doc) > 0 && options::show_user_defined_docstrings()) {
                 // If we're appending another docstring, and aren't printing function signatures, we
                 // need to append a newline first:
-                if (!options::show_function_signatures()) {
-                    if (first_user_def) first_user_def = false;
-                    else signatures += "\n";
+                if (!show_signature_headings && first_user_def) {
+                     first_user_def = false;
+                }
+                else {
+                    signatures += "\n";
                 }
-                if (options::show_function_signatures()) signatures += "\n";
                 signatures += it->doc;
-                if (options::show_function_signatures()) signatures += "\n";
+                if (show_signature_headings) {
+                    signatures += "\n";
```

Comment:
If you're really feeling adventurous: it seems there's always a newline attached to the end of the docstring if `show_signature_headings`? Not worth making the code more complicate for, but maybe this helps restructuring the logic?

If you want, I can have a look, as well.

Reply:
The original code always added a new line. But if we wanted to get rid of it then I think it would make the code more complicated. I'm thinking we would have to do something like

```cpp
        for (auto it = chain_start; it != nullptr; it = it->next) {
            bool appending_docstring = (it->doc && strlen(it->doc) > 0 && options::show_user_defined_docstrings());
            bool last_section = (it->next == nullptr);
            if (show_signature_headings) {
                if (index > 0) signatures += "\n";
                if (chain)
                    signatures += std::to_string(++index) + ". ";
                signatures += rec->name;
                signatures += it->signature;
                if (appending_docstring || !last_section)) signatures += "\n";
            }
            if (appending_docstring) {
                // If we're appending another docstring, and aren't printing function signatures, we
                // need to append a newline first:
                if (!show_signature_headings && first_user_def)
                     first_user_def = false;
                else
                    signatures += "\n";
                signatures += it->doc;
                if (show_signature_headings && !last_section) signatures += "\n";
            }
        }
```

Reply:
Your call on whether it's worth it. I know the original code did it, but it seems rather silly. I had hoped it would make the code easier/more intuitive. :-(